### DISCUSSION TITLE
(G-CLI 3) -1 Exit not working?

### DISCUSSION THREAD

**Discussion by HunterJKI** [JamesMc86/G-CLI#148] [General]
I am using G-CLI 3.0.1.98

The VI I am calling into is causing a LabVIEW crash (not related to g-cli, I think), in the console G-CLI shows it loses connection and exits. It states it exists with error code -1, but it appears to not do that.

This test is being run inside a bitbucket pipline, and the fact that the pipeline is not failing is my evidence that it is not returning, It fails the pipeline when I exit with an error, and I verified Bitbucket fails with a -1 error code.

```
18:44:19.347 [ERROR] Comms Error: IO Error When Reading Messages From LabVIEW
18:44:19.347 [ERROR] Comms Error: IO Error When Reading Messages From LabVIEW
18:44:19.347 [DEBUG] Comms reader stopped.
18:44:19.378 [DEBUG] Signal Handler Stopped.
18:44:19.378 [DEBUG] Action loop stopped.
18:44:19.378 [DEBUG] Waiting on monitoring thread to complete
18:44:20.018 [DEBUG] Monitoring complete and kill not requested.
18:44:20.018 [DEBUG] Stopping LabVIEW monitoring due to stop command from application
18:44:20.018 [DEBUG] Monitoring thread completed
18:44:20.018 [DEBUG] Exiting G-CLI with exit code -1
```

I reverted to G-CLI 2.4, when my code crashes, G-CLI does cause the job to fail. Is there an exit code bug? 

There are lots of variables in the mix, so I am not yet confident this is a bug, just trying to narrow down the root cause.

**Comment by JamesWiresmith** [comment]
Let me take a look. I don't see any obvious place in the code it is dropped. What OS/shell are you running this in? (I assume windows but powershell and cmd can act differently in pipelines)

**Reply by HunterJKI** [reply]
Its using powershell.

**Comment by JamesWiresmith** [comment]
Great I'll do some testing. Powershell generally ignores exit codes but the Comms error lines should be on StdErr which powershell will normally take as an error.

The new version was supposed to be more powershell friendly though so looks like I've missed something

James McNally
Director/Systems Engineer
Wiresmith Technology
t: 01635 597171
________________________________

**Reply by HunterJKI** [reply]
@JamesMc86 Have you been able to reproduce this? We keep seeing it.

**Comment by JamesWiresmith** [comment]
This week got away from me but I have some time set aside next week to work on G CLI

James McNally
Director/Systems Engineer
Wiresmith Technology
t: 01635 597171
________________________________

**Comment by JamesWiresmith** [comment]
I've been trying to track down any specific information for bitbucket and struggling so I'm going to give some general things I know with powershell and you may need to try some things in bitbucket.

G-CLI indicates errors in two ways:

1. Output to StdErr
2. Return non-zero exit code

2 is actually ignored by default in powershell. 1 is detected but by default is a non-terminating error. This means powershell will merrily carry on by default.

In your log, the top two lines are of the error type so should be going to standard error.

In most powershell environments you have to change these non-terminating errors to terminating errors in some way. In the integration tests I do it by setting $ErrorActionPreference to stop at https://github.com/JamesMc86/G-CLI/blob/e68f1f85d9b40c5c44cc53fb163c73ce207920cb/run_integration_tests.ps1#L2. I think you may need the same or similar for bitbucket.

This response on StackOverflow shows the progression of adding error handling in powershell for native apps: https://stackoverflow.com/questions/65734362/handle-error-from-external-console-application-in-powershell-script

In short though error handling from native apps in powershell seems to be a mess unless I'm missing something. I'd happily add specific support to G-CLI if I can find any clear way to do it.

Hopefully that is a starting point and happy to dig further with you. But I'm pretty sure your hitting against something like this.

**Comment by JamesWiresmith** [comment]
Here is a one-liner that does throw an error in powershell (but not pwsh):

`$errorActionPreference="stop"; try { g-cli ".\Generate Large Error.vi" -- 10  2>null } catch { throw "Error Detected" }`

The redirection seems to be important (which is documented in a few places)

But I think checking the return code as in the StackOverflow above is probably the most reliable method.

I've also created this issue to see if we can create a cmdlet to make this more reliable: https://github.com/JamesMc86/G-CLI/issues/149
